perf/light.cc

```diff
@@ -0,0 +1,653 @@
+#include <algorithm>
+#include <climits>
+#include <cstring>
+#include <random>
+#include <vector>
+#include <unordered_set>
+
+#include <benchmark/benchmark.h>
+
+// This test contains benchmarks for Light - data structure implementing
+// Tarantool HASH index and being hash table under the hood.
+//
+// Let's use raw payload (similar to native Tarantool tuples) as values.
+// Test scenarios:
+//  - Inserts only;
+//  - Search only (by value), no misses;
+//  - Search only (by value) with misses;
+//  - Search by key;
+//  - Sequence iteration;
+//  - Inserts after erase;
+//  - Inserts alongside with lookups;
+//  - Search after erase;
+//  - Deletes.
+//
+// To compare numbers given by this benchmark we also run a few tests with
+// std::unordered_map at the end.
+
+
+// Tuple size in fact does not really matter since in Tarantool we store
+// pointers to them (since light hash table does not support entries with
+// sizeof >= 8 bytes). So let's use just random one :)
+constexpr static std::size_t TUPLE_SIZE = 1 << 5;
+
+// We measure perf on large datasets: min number of elements is
+// considered to be 10k; max - 1M (which I guess covers the most
+// popular amount of data stored in Tarantool indexes).
+constexpr static std::size_t TUPLE_COUNT_MIN = 10000;
+constexpr static std::size_t TUPLE_COUNT_MAX = 100 * TUPLE_COUNT_MIN;
+constexpr static std::size_t TUPLE_COUNT_MULTIPLIER = 10;
+
+////////////////////////////// Data Definitions ////////////////////////////////////////////////////////////////////////
+
+// 'size' can be skipped in fact (since it is the same for all tuples).
+// Let's keep it just in case (and to make benchmark look a bit more closer
+// to real world). In order to compare two TupleRaw we use memcpy.
+struct TupleRaw {
+	TupleRaw(unsigned char *ptr, std::size_t len) :
+		data(ptr), size(len) { }
+	unsigned char *data;
+	std::size_t size;
+};
+
+using Key_t   = uint8_t;
+using Hash_t  = uint32_t;
+using Value_t = TupleRaw;
+
+// Light hash map implementation does not include any hashing function;
+// it's up to user to choose the proper one. Let's use quite primitive
+// yet good enough: Fowler–Noll–Vo (-1a).
+namespace Hash {
```

Comment:
Why don't you use MurmurHash3, which is used for tuple hashing natively?

Reply:
Because we make benchmark for light disregarding on hash function?

Reply:
@alyapunov for this reason as well.

Reply:
Discussed f2f, I guess we agreed... tl;dr to make bench closer to real world.